src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -277,21 +287,355 @@ openLevels reg hfs hbio TableConfig{..} sessionRoot levels =
+      <> encodeWord (minor ver)
+
+instance Decode SnapshotVersion where
+  decode = do
+      _ <- decodeListLenOf 2
+      majo <- decodeWord
+      mino <- decodeWord
+      case fromMajorMinor majo mino of
+        Nothing -> fail ("Unknown version number: " <> prettyVersion majo mino)
+        Just version -> pure version
+
+{-------------------------------------------------------------------------------
+  Encoding and decoding: SnapshotMetaData
+-------------------------------------------------------------------------------}
+
+-- SnapshotMetaData
+
+instance Encode SnapshotMetaData where
+  encode (SnapshotMetaData label config levels) =
+         encodeListLen 3
+      <> encode label
+      <> encode config
+      <> encode levels
+
+instance DecodeVersioned SnapshotMetaData where
+  decode' ver@V0_0 = do
+      _ <- decodeListLenOf 3
+      SnapshotMetaData <$> decode' ver <*> decode' ver <*> decode' ver
+
+-- SnapshotLabel
+
+instance Encode SnapshotLabel where
+  encode (SnapshotLabel s) = encodeString s
+
+instance DecodeVersioned SnapshotLabel where
+  decode' V0_0 = SnapshotLabel <$> decodeString
+
+{-------------------------------------------------------------------------------
+  Encoding and decoding: TableConfig
 -------------------------------------------------------------------------------}
 
-deriving stock instance Read MergePolicyForLevel
-deriving newtype instance Read RunNumber
-deriving stock instance Read Merge.Level
+-- TableConfig
+
+instance Encode TableConfig where
+  encode config =
+         encodeListLen 7
+      <> encode mergePolicy
+      <> encode sizeRatio
+      <> encode writeBufferAlloc
+      <> encode bloomFilterAlloc
+      <> encode fencePointerIndex
+      <> encode diskCachePolicy
+      <> encode mergeSchedule
+    where
+      TableConfig
+        mergePolicy
+        sizeRatio
+        writeBufferAlloc
+        bloomFilterAlloc
+        fencePointerIndex
+        diskCachePolicy
+        mergeSchedule
+        = config
+
+instance DecodeVersioned TableConfig where
+  decode' v@V0_0 = do
+      _ <- decodeListLenOf 7
+      TableConfig
+        <$> decode' v <*> decode' v <*> decode' v
+        <*> decode' v <*> decode' v <*> decode' v
+        <*> decode' v
+
+-- MergePolicy
+
+instance Encode MergePolicy where
+  encode MergePolicyLazyLevelling = encodeWord 0
+
+instance DecodeVersioned MergePolicy where
+  decode' V0_0 =  do
+      tag <- decodeWord
+      case tag of
+        0 -> pure MergePolicyLazyLevelling
+        _ -> fail ("Unexpected tag: " <> show tag)
+
+-- SizeRatio
+
+instance Encode SizeRatio where
+  encode Four = encodeWord64 4
+
+instance DecodeVersioned SizeRatio where
+  decode' V0_0 = do
+      x <- decodeWord64
+      case x of
+        4 -> pure Four
+        _ -> fail ("Expected 4, but found " <> show x)
+
+-- WriteBufferAlloc
+
+instance Encode WriteBufferAlloc where
+  encode (AllocNumEntries numEntries) =
+         encodeTag 0
+      <> encode numEntries
```

Comment:
We use `encodeWord` for all other sum types. Is there a relevant difference? Should we use `encodeTag` there, too?

Reply:
Probably a mistake. I initially used `encodeTag`/`decodeTag` everywhere, but "tags" are their own specific concept in CBOR. Because of the misuse, the `validFlatTerm` property would fail, so I changed it all to use `encodeWord`/`decodeWord`.

FYI, CBOR tags describe the type of the CBOR item, such as it being a byte string, text string, array, etc.